## 1. Importing the libraries

In [3]:
import os
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split

# 2. Reading the Data

In [4]:
PROJECT_DIR = r"C:\Users\jaragst\OneDrive - Pella Corporation\ML\Flight-Prediction-Sagemaker"
DATA_DIR = "data"

In [5]:
def get_data(name):
    file_name = f"{name}.csv"
    file_path = os.path.join(PROJECT_DIR,DATA_DIR,file_name)
    return pd.read_csv(file_path)

In [6]:
flights = get_data("flight_price")
flights

Airline Date_of_Journey    Source Destination  \
0           IndiGo      24/03/2019  Banglore   New Delhi   
1        Air India       1/05/2019   Kolkata    Banglore   
2      Jet Airways       9/06/2019     Delhi      Cochin   
3           IndiGo      12/05/2019   Kolkata    Banglore   
4           IndiGo      01/03/2019  Banglore   New Delhi   
...            ...             ...       ...         ...   
10678     Air Asia       9/04/2019   Kolkata    Banglore   
10679    Air India      27/04/2019   Kolkata    Banglore   
10680  Jet Airways      27/04/2019  Banglore       Delhi   
10681      Vistara      01/03/2019  Banglore   New Delhi   
10682    Air India       9/05/2019     Delhi      Cochin   

                       Route Dep_Time  Arrival_Time Duration Total_Stops  \
0                  BLR → DEL    22:20  01:10 22 Mar   2h 50m    non-stop   
1      CCU → IXR → BBI → BLR    05:50         13:15   7h 25m     2 stops   
2      DEL → LKO → BOM → COK    09:25  04:25 10 Jun      19h     2 stops   
3            CCU → NAG → BLR    18:05         23:30   5h 25m      1 stop   
4            BLR → NAG → DEL    16:50         21:35   4h 45m      1 stop   
...                      ...      ...           ...      ...         ...   
10678              CCU → BLR    19:55         22:25   2h 30m    non-stop   
10679              CCU → BLR    20:45         23:20   2h 35m    non-stop   
10680              BLR → DEL    08:20         11:20       3h    non-stop   
10681              BLR → DEL    11:30         14:10   2h 40m    non-stop   
10682  DEL → GOI → BOM → COK    10:55         19:15   8h 20m     2 stops   

      Additional_Info  Price  
0             No info   3897  
1             No info   7662  
2             No info  13882  
3             No info   6218  
4             No info  13302  
...               ...    ...  
10678         No info   4107  
10679         No info   4145  
10680         No info   7229  
10681         No info  12648  
10682         No info  11753  

[10683 rows x 11 columns]

In [7]:
flights.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10683 non-null  object
 1   Date_of_Journey  10683 non-null  object
 2   Source           10683 non-null  object
 3   Destination      10683 non-null  object
 4   Route            10682 non-null  object
 5   Dep_Time         10683 non-null  object
 6   Arrival_Time     10683 non-null  object
 7   Duration         10683 non-null  object
 8   Total_Stops      10682 non-null  object
 9   Additional_Info  10683 non-null  object
 10  Price            10683 non-null  int64 
dtypes: int64(1), object(10)
memory usage: 918.2+ KB


- The dataset contains 10,683 rows and 11 features
- Columns 'Route' and 'Total_Stops' have missing value each
- The data types of some features isn't appropriate
    

# 3. Preliminary Analysis

## 3.1 Check Data Types

In [8]:
flights.head()

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897  
1    05:50         13:15   7h 25m     2 stops         No info   7662  
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882  
3    18:05         23:30   5h 25m      1 stop         No info   6218  
4    16:50         21:35   4h 45m      1 stop         No info  13302

In [9]:
flights.dtypes

Airline            object
Date_of_Journey    object
Source             object
Destination        object
Route              object
Dep_Time           object
Arrival_Time       object
Duration           object
Total_Stops        object
Additional_Info    object
Price               int64
dtype: object

In [10]:
value = flights.Date_of_Journey.iloc[6]
value

'12/03/2019'

In [11]:
value = flights.Dep_Time.iloc[6]
value

'18:55'

## 3.2 Check for Duplicates

In [12]:
flights.duplicated().sum()

np.int64(220)

In [13]:
(
    flights.
    loc[flights.duplicated(keep=False)]
    .sort_values(["Airline","Date_of_Journey","Source","Destination"])
)

Airline Date_of_Journey    Source Destination                  Route  \
6321  Air India      01/03/2019  Banglore   New Delhi  BLR → BOM → AMD → DEL   
9848  Air India      01/03/2019  Banglore   New Delhi  BLR → BOM → AMD → DEL   
572   Air India      03/03/2019  Banglore   New Delhi              BLR → DEL   
8168  Air India      03/03/2019  Banglore   New Delhi              BLR → DEL   
1495  Air India       1/04/2019   Kolkata    Banglore  CCU → DEL → COK → BLR   
...         ...             ...       ...         ...                    ...   
2692   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
2870   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
3711   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
2634    Vistara      24/03/2019  Banglore   New Delhi              BLR → DEL   
5470    Vistara      24/03/2019  Banglore   New Delhi              BLR → DEL   

     Dep_Time  Arrival_Time Duration Total_Stops  \
6321    08:50  23:55 02 Mar   39h 5m     2 stops   
9848    08:50  23:55 02 Mar   39h 5m     2 stops   
572     21:10         23:55   2h 45m    non-stop   
8168    21:10         23:55   2h 45m    non-stop   
1495    10:00  01:20 02 Apr  15h 20m     2 stops   
...       ...           ...      ...         ...   
2692    05:45         08:35   2h 50m    non-stop   
2870    05:45         08:35   2h 50m    non-stop   
3711    20:30         23:20   2h 50m    non-stop   
2634    11:30         14:10   2h 40m    non-stop   
5470    11:30         14:10   2h 40m    non-stop   

                   Additional_Info  Price  
6321                       No info  17135  
9848                       No info  17135  
572                        No info   7591  
8168                       No info   7591  
1495                       No info  10408  
...                            ...    ...  
2692  No check-in baggage included   4273  
2870  No check-in baggage included   4273  
3711  No check-in baggage included   3873  
2634                       No info   5403  
5470                       No info   5403  

[415 rows x 11 columns]

## 3.3 Observations

- The type of ' Date_of_Journey','Dep_Time' should be changes to datetime

- The type of ' Duration' is mixed .It should be numeric value.Can be in minutes

- 'Total_Stops' should be in numeric.Should not contain string after that

- There are total 220 values are duplicates which need to be removed

## 4. Detailed Analysis

### Airline

flights.Airline

In [14]:
flights.Airline.unique()

array(['IndiGo', 'Air India', 'Jet Airways', 'SpiceJet',
       'Multiple carriers', 'GoAir', 'Vistara', 'Air Asia',
       'Vistara Premium economy', 'Jet Airways Business',
       'Multiple carriers Premium economy', 'Trujet'], dtype=object)

- Some of the entries have inconsistance /inaccurate values

In [15]:
(
    flights
    .Airline
    .str.replace(" Premium economy","")
    .str.replace(" Airways Business","")
    .str.title()
    .unique()
)

array(['Indigo', 'Air India', 'Jet Airways', 'Spicejet',
       'Multiple Carriers', 'Goair', 'Vistara', 'Air Asia', 'Jet',
       'Trujet'], dtype=object)

#### Date Of Journey

In [16]:
flights.Date_of_Journey

0        24/03/2019
1         1/05/2019
2         9/06/2019
3        12/05/2019
4        01/03/2019
            ...    
10678     9/04/2019
10679    27/04/2019
10680    27/04/2019
10681    01/03/2019
10682     9/05/2019
Name: Date_of_Journey, Length: 10683, dtype: object

In [17]:
pd.to_datetime(flights.Date_of_Journey,dayfirst=True)

0       2019-03-24
1       2019-05-01
2       2019-06-09
3       2019-05-12
4       2019-03-01
           ...    
10678   2019-04-09
10679   2019-04-27
10680   2019-04-27
10681   2019-03-01
10682   2019-05-09
Name: Date_of_Journey, Length: 10683, dtype: datetime64[ns]

#### Source

In [18]:
flights.Source.unique()

array(['Banglore', 'Kolkata', 'Delhi', 'Chennai', 'Mumbai'], dtype=object)

#### Destination

In [19]:
flights.Destination.unique()

array(['New Delhi', 'Banglore', 'Cochin', 'Kolkata', 'Delhi', 'Hyderabad'],
      dtype=object)

#### Route

- we dont need this column as we have source and  destination already to capture info

#### Departure Time

In [20]:
flights.Dep_Time

0        22:20
1        05:50
2        09:25
3        18:05
4        16:50
         ...  
10678    19:55
10679    20:45
10680    08:20
10681    11:30
10682    10:55
Name: Dep_Time, Length: 10683, dtype: object

###### Check if column have other than numbers

In [21]:
(
    flights
    .Dep_Time
    .loc[lambda ser : ser.str.contains("[^0-9:]")]
)

Series([], Name: Dep_Time, dtype: object)

- There is no string or special char in column 'Dep_Time'

In [22]:
pd.to_datetime(flights.Dep_Time).dt.time

C:\Users\jaragst\AppData\Local\Temp\ipykernel_9136\1056054517.py:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  pd.to_datetime(flights.Dep_Time).dt.time


0        22:20:00
1        05:50:00
2        09:25:00
3        18:05:00
4        16:50:00
           ...   
10678    19:55:00
10679    20:45:00
10680    08:20:00
10681    11:30:00
10682    10:55:00
Name: Dep_Time, Length: 10683, dtype: object

#### Arrival Time 

In [23]:
flights.Arrival_Time

0        01:10 22 Mar
1               13:15
2        04:25 10 Jun
3               23:30
4               21:35
             ...     
10678           22:25
10679           23:20
10680           11:20
10681           14:10
10682           19:15
Name: Arrival_Time, Length: 10683, dtype: object

In [24]:
(
    flights
    .Arrival_Time
    .loc[lambda ser: ser.str.contains("[^0-9:]")]
    .str.split(" ",n =1)
    .str.get(1)
    .unique()
)
   

array(['22 Mar', '10 Jun', '13 Mar', '02 Mar', '10 May', '04 Mar',
       '13 Jun', '28 May', '19 Mar', '07 May', '02 Jun', '16 Jun',
       '19 May', '16 May', '28 Jun', '02 May', '28 Mar', '19 Jun',
       '04 Apr', '25 Mar', '07 Mar', '25 Jun', '07 Jun', '25 May',
       '13 May', '16 Mar', '22 May', '10 Apr', '04 Jun', '20 May',
       '28 Apr', '25 Apr', '10 Mar', '19 Apr', '13 Apr', '02 Apr',
       '23 Mar', '22 Apr', '11 May', '07 Apr', '03 May', '08 Mar',
       '03 Mar', '05 Mar', '22 Jun', '04 May', '26 May', '16 Apr',
       '26 Jun', '29 May', '29 Jun', '29 Mar', '23 May', '17 Jun'],
      dtype=object)

#### Duration

In [25]:
flights.Duration

0        2h 50m
1        7h 25m
2           19h
3        5h 25m
4        4h 45m
          ...  
10678    2h 30m
10679    2h 35m
10680        3h
10681    2h 40m
10682    8h 20m
Name: Duration, Length: 10683, dtype: object

In [26]:
(
    flights.
    Duration
    .loc[lambda ser : ~ser.str.contains("m")]
    .unique()
)

array(['19h', '23h', '22h', '12h', '3h', '5h', '10h', '18h', '24h', '15h',
       '16h', '8h', '14h', '20h', '13h', '11h', '9h', '27h', '26h', '4h',
       '7h', '30h', '21h', '28h', '47h', '6h', '25h', '38h', '34h'],
      dtype=object)

In [27]:
(
    flights.
    Duration
    .loc[lambda ser : ~ser.str.contains("h")]
   # .unique()
)

6474    5m
Name: Duration, dtype: object

In [28]:
flights.iloc[[6474]]

Airline Date_of_Journey  Source Destination                  Route  \
6474  Air India       6/03/2019  Mumbai   Hyderabad  BOM → GOI → PNQ → HYD   

     Dep_Time Arrival_Time Duration Total_Stops Additional_Info  Price  
6474    16:50        16:55       5m     2 stops         No info  17327

 - The observation indexed 6474 has duration 5 minutes so we will  remove the value

In [29]:
(
    flights
    .Duration
    .drop(index =6474)
    .str.split(" ",expand = True)
    .set_axis(["hour","minute"],axis =1)
    .assign(
        hour = lambda df_: (
            df_
            .hour
            .str.replace("h","")
            .astype(int)
            .mul(60)
        ),
        minute = lambda df_ :(
            df_
            .minute
            .str.replace("m","")
            .fillna("0")
            .astype(int)
        )
        
        
    )
    .sum(axis=1)
    .rename("duration_minutes")
    .to_frame()
    .join(flights.Duration.drop(index=6474))
    #.dtypes
    #.isna().sum()
)

duration_minutes Duration
0                   170   2h 50m
1                   445   7h 25m
2                  1140      19h
3                   325   5h 25m
4                   285   4h 45m
...                 ...      ...
10678               150   2h 30m
10679               155   2h 35m
10680               180       3h
10681               160   2h 40m
10682               500   8h 20m

[10682 rows x 2 columns]

In [30]:
(
    flights
    .Duration
    .drop(index =6474)
    .str.split(" ",expand = True)
    .set_axis(["hour","minute"],axis =1)
    .assign(
        hour = lambda df_: (
            df_
            .hour
            .str.replace("h","")
            .astype(int)
            .mul(60)
        ),
        minute = lambda df_ :(
            df_
            .minute
            .str.replace("m","")
            .fillna("0")
            .astype(int)
        )
        
        
    )
    .sum(axis=1)
    #.dtypes
    #.isna().sum()
)

0         170
1         445
2        1140
3         325
4         285
         ... 
10678     150
10679     155
10680     180
10681     160
10682     500
Length: 10682, dtype: int64

#### Total Stops

In [31]:
flights.Total_Stops.unique()

array(['non-stop', '2 stops', '1 stop', '3 stops', nan, '4 stops'],
      dtype=object)

In [32]:
(
    flights
    .Total_Stops
    .replace("non-stop","0")
    .str.replace(" stops?", "",regex=True)#"stop" or "stops" s is optional
    #.astype(int) will not work as missing values
    .pipe(lambda ser : pd.to_numeric(ser))
)

0        0.0
1        2.0
2        2.0
3        1.0
4        1.0
        ... 
10678    0.0
10679    0.0
10680    0.0
10681    0.0
10682    2.0
Name: Total_Stops, Length: 10683, dtype: float64

#### Additinal Info

In [33]:
flights.Additional_Info.unique()

array(['No info', 'In-flight meal not included',
       'No check-in baggage included', '1 Short layover', 'No Info',
       '1 Long layover', 'Change airports', 'Business class',
       'Red-eye flight', '2 Long layover'], dtype=object)

## 5.Cleaning Operations

In [34]:
def convert_to_minutes(ser):
    return(
        ser
        .str.split(" ",expand = True)
        .set_axis(["hour","minute"],axis =1)
        .assign(
            hour = lambda df_: (
            df_
            .hour
            .str.replace("h","")
            .astype(int)
            .mul(60)
                ),
            minute = lambda df_ :(
            df_
            .minute
            .str.replace("m","")
            .fillna("0")
            .astype(int)
            )    
    )
    .sum(axis=1)
   # .rename("duration_minutes")
    #.to_frame()
    #.join(flights.Duration.drop(index=6474))
        
    )

In [41]:
def clean_data(df):
    return(
        df
        .drop(index=6474)
        .drop_duplicates()
        .assign(**{
            col : df[col].str.strip()
            for col in df.select_dtypes(include = "O").columns
           })
        .rename(columns=str.lower)
        .assign(
            airline = lambda df_:(
                df_
                .airline
                .str.replace(" Premium economy","")
                .str.replace(" Airways Business","")
                .str.title()
             ),
            date_of_journey = lambda df_:pd.to_datetime(df_.date_of_journey,dayfirst=True),
            dep_time =lambda df_:pd.to_datetime(df_.dep_time).dt.time,
            arrival_time =lambda df_:pd.to_datetime(df_.arrival_time).dt.time,
            duration = lambda df_:df_.duration.pipe(convert_to_minutes),
            total_stops = lambda df_ :(
                       df_
                       .total_stops
                       .replace("non-stop","0")
                       .str.replace(" stops?", "",regex=True)#"stop" or "stops" s is optional
                            #.astype(int) will not work as missing values
                       .pipe(lambda ser : pd.to_numeric(ser))
      ),
            additional_info = lambda df_:df_.additional_info.replace("No info","No Info")
            )
        .drop(columns = "route")
    )  
        

In [42]:
clean_data(flights)#.dtypes#.airline.unique()

C:\Users\jaragst\AppData\Local\Temp\ipykernel_9136\2584866168.py:20: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  dep_time =lambda df_:pd.to_datetime(df_.dep_time).dt.time,
C:\Users\jaragst\AppData\Local\Temp\ipykernel_9136\2584866168.py:21: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  arrival_time =lambda df_:pd.to_datetime(df_.arrival_time).dt.time,


airline date_of_journey    source destination  dep_time  \
0           Indigo      2019-03-24  Banglore   New Delhi  22:20:00   
1        Air India      2019-05-01   Kolkata    Banglore  05:50:00   
2      Jet Airways      2019-06-09     Delhi      Cochin  09:25:00   
3           Indigo      2019-05-12   Kolkata    Banglore  18:05:00   
4           Indigo      2019-03-01  Banglore   New Delhi  16:50:00   
...            ...             ...       ...         ...       ...   
10678     Air Asia      2019-04-09   Kolkata    Banglore  19:55:00   
10679    Air India      2019-04-27   Kolkata    Banglore  20:45:00   
10680  Jet Airways      2019-04-27  Banglore       Delhi  08:20:00   
10681      Vistara      2019-03-01  Banglore   New Delhi  11:30:00   
10682    Air India      2019-05-09     Delhi      Cochin  10:55:00   

      arrival_time  duration  total_stops additional_info  price  
0         01:10:00       170          0.0         No Info   3897  
1         13:15:00       445          2.0         No Info   7662  
2         04:25:00      1140          2.0         No Info  13882  
3         23:30:00       325          1.0         No Info   6218  
4         21:35:00       285          1.0         No Info  13302  
...            ...       ...          ...             ...    ...  
10678     22:25:00       150          0.0         No Info   4107  
10679     23:20:00       155          0.0         No Info   4145  
10680     11:20:00       180          0.0         No Info   7229  
10681     14:10:00       160          0.0         No Info  12648  
10682     19:15:00       500          2.0         No Info  11753  

[10462 rows x 10 columns]

#### dictionary unpacking 

In [ ]:
def print_person(name, age, city):
    print(f"{name} is {age} years old and lives in {city}")
    
person = {'name': 'John', 'age': 30, 'city': 'New York'}
print_person(**person)

In [47]:
flight_cleaned = clean_data(flights)
flight_cleaned

C:\Users\jaragst\AppData\Local\Temp\ipykernel_9136\2584866168.py:20: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  dep_time =lambda df_:pd.to_datetime(df_.dep_time).dt.time,
C:\Users\jaragst\AppData\Local\Temp\ipykernel_9136\2584866168.py:21: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  arrival_time =lambda df_:pd.to_datetime(df_.arrival_time).dt.time,


airline date_of_journey    source destination  dep_time  \
0           Indigo      2019-03-24  Banglore   New Delhi  22:20:00   
1        Air India      2019-05-01   Kolkata    Banglore  05:50:00   
2      Jet Airways      2019-06-09     Delhi      Cochin  09:25:00   
3           Indigo      2019-05-12   Kolkata    Banglore  18:05:00   
4           Indigo      2019-03-01  Banglore   New Delhi  16:50:00   
...            ...             ...       ...         ...       ...   
10678     Air Asia      2019-04-09   Kolkata    Banglore  19:55:00   
10679    Air India      2019-04-27   Kolkata    Banglore  20:45:00   
10680  Jet Airways      2019-04-27  Banglore       Delhi  08:20:00   
10681      Vistara      2019-03-01  Banglore   New Delhi  11:30:00   
10682    Air India      2019-05-09     Delhi      Cochin  10:55:00   

      arrival_time  duration  total_stops additional_info  price  
0         01:10:00       170          0.0         No Info   3897  
1         13:15:00       445          2.0         No Info   7662  
2         04:25:00      1140          2.0         No Info  13882  
3         23:30:00       325          1.0         No Info   6218  
4         21:35:00       285          1.0         No Info  13302  
...            ...       ...          ...             ...    ...  
10678     22:25:00       150          0.0         No Info   4107  
10679     23:20:00       155          0.0         No Info   4145  
10680     11:20:00       180          0.0         No Info   7229  
10681     14:10:00       160          0.0         No Info  12648  
10682     19:15:00       500          2.0         No Info  11753  

[10462 rows x 10 columns]

## 6. Split The Data

In [53]:
flights_final = flight_cleaned.sample(1000)

In [54]:
X = flights_final.drop(columns = "price")
y = flights_final.price.copy()

In [55]:
X_,X_test,y_,y_test = train_test_split(X,y,test_size = 0.2,random_state =42)
X_train,X_val,y_train,y_val = train_test_split(X_,y_,test_size = 0.2,random_state=42)

In [56]:
print(X_train.shape,y_train.shape)
print(X_val.shape,y_val.shape)
print(X_test.shape,y_test.shape)

(640, 9) (640,)
(160, 9) (160,)
(200, 9) (200,)


## 7. Export the subsets

In [60]:
def export_data(X,y,name):
    file_name = f"{name}.csv"
    file_path = os.path.join(PROJECT_DIR,DATA_DIR,file_name)
    X.join(y).to_csv(file_path,index=False)
    return pd.read_csv(file_path).head()

In [61]:
export_data(X_train,y_train,"train")

airline date_of_journey    source destination  dep_time arrival_time  \
0       Indigo      2019-04-01     Delhi      Cochin  06:50:00     12:10:00   
1  Jet Airways      2019-05-27     Delhi      Cochin  15:00:00     04:25:00   
2    Air India      2019-03-01  Banglore   New Delhi  11:05:00     22:45:00   
3  Jet Airways      2019-05-24   Kolkata    Banglore  18:55:00     19:50:00   
4  Jet Airways      2019-04-01     Delhi      Cochin  16:00:00     12:35:00   

   duration  total_stops              additional_info  price  
0       320          1.0                      No Info   5298  
1       805          1.0  In-flight meal not included  12898  
2       700          1.0                      No Info  14069  
3      1495          1.0  In-flight meal not included  10844  
4      1235          1.0  In-flight meal not included   5406

In [62]:
export_data(X_test,y_test,"test")

airline date_of_journey    source destination  dep_time arrival_time  \
0  Jet Airways      2019-06-27     Delhi      Cochin  09:00:00     12:35:00   
1  Jet Airways      2019-03-03    Mumbai   Hyderabad  02:55:00     04:25:00   
2     Spicejet      2019-03-09   Kolkata    Banglore  09:00:00     11:30:00   
3     Spicejet      2019-04-03    Mumbai   Hyderabad  22:45:00     00:15:00   
4      Vistara      2019-03-09  Banglore   New Delhi  09:50:00     12:35:00   

   duration  total_stops              additional_info  price  
0      1655          1.0  In-flight meal not included  10262  
1        90          0.0  In-flight meal not included   4050  
2       150          0.0                      No Info   5515  
3        90          0.0                      No Info   1759  
4       165          0.0                      No Info  11598

In [63]:
export_data(X_val,y_val,"val")

airline date_of_journey   source destination  dep_time arrival_time  \
0    Air India      2019-04-01  Kolkata    Banglore  10:00:00     10:30:00   
1       Indigo      2019-05-15    Delhi      Cochin  06:50:00     12:10:00   
2     Spicejet      2019-04-06  Kolkata    Banglore  17:10:00     19:40:00   
3  Jet Airways      2019-05-12  Kolkata    Banglore  18:55:00     08:15:00   
4       Indigo      2019-06-15   Mumbai   Hyderabad  16:55:00     18:25:00   

   duration  total_stops               additional_info  price  
0      1470          3.0                       No Info  12954  
1       320          1.0                       No Info   5298  
2       150          0.0  No check-in baggage included   3841  
3       800          1.0   In-flight meal not included   9663  
4        90          0.0                       No Info   3175